File: 89-gray-code/89-gray-code.py
class Solution:
    def grayCode(self, n: int) -> List[int]:
        def generate(n):
            if n == 1:
                return [0,1]
            prev = generate(n-1)
            temp = []
            for i in range(len(prev)-1,-1,-1):
                temp.append(prev[i] + 2**(n-1))
            return prev + temp
        
        return generate(n)       
        
            
        
            

File: 89-gray-code/89-gray-code.java
class Solution {
    public List<Integer> grayCode(int n) {
        //recursion
        //iteration
        return generate(n);
        
        
    }
    public ArrayList<Integer> generate(int n){
        if (n == 1){
            return new ArrayList<Integer>(Arrays.asList(0,1));
        }
        ArrayList<Integer> prev = generate(n - 1);
        int mask = 1<<(n-1);
        for(int i = prev.size() - 1; i >= 0; i--){
            int e = prev.get(i);
            prev.add(e|mask);
        }
        return prev;
    }
    
    
}